### DISCUSSION TITLE
Trying to make a generic recursive node list

### DISCUSSION THREAD

**Discussion by danielo515** [L3MON4D3/LuaSnip#689] [Q&A]
Hello, inspired by the infinite recursive snippet from the wiki, I tried to make a generic version that I can use like a library in all my snippets. however, I'm failing to achieve that goal. Maybe you can take a look at my code and tell me what is wrong.
Thanks!

here is the code:

```lua
local rec_ls
rec_ls = function(pos, _nodes)
  assert(type(_nodes) == "table", "List of nodes must be a table")
  local nodes = vim.list_extend({ t "" }, _nodes)
  return function()
    table.insert(nodes, d(pos, rec_ls(pos, _nodes), {}))
    vim.pretty_print(nodes)
    return sn(nil, {
      c(1, {
        -- important!! Having the sn(...) as the first choice will cause infinite recursion.
        t { "" },
        -- The same dynamicNode as in the snippet (also note: self reference).
        sn(nil, nodes),
      }),
    })
  end
end
```

The used like this:
```lua
  s(
    { trig = "ext", dscr = "Extern class", regTrig = false },

    fmt(
      [[ extern class {} {{
              {}
              {}
              }}
    ]],
      {
        i(1),
        d(2, rec_ls(3, fmt(" static function {}({}):{}; ", { i(1), i(2), i(0) })), {}),
      }
    )
  ),
```

**Comment by L3MON4D3** [comment]
I think tye issue is that the `_nodes` are reused in the different "stages" of t€hat snippet. Try passing `_nodes:copy()` to the recursive call instead.
But also generally, I'd advise you to follow the warning in the wiki, use that dynamicNode-extension, it is a bit more complicated, but worth it imo :D

**Reply by danielo515** [reply]
Copying the nodes is the first thing is done at the top of the function (see the list_extend call). I will re-read the wiki article, because I don't remember any reference to any other method there.

**Reply by L3MON4D3** [reply]
Oh, I was referring to a deep copy, with a shallow copy (which afaik list_extend performs) there might still be the same exact node multiple times inside the snippet, which can cause all sorts of issues. `local nodes = _nodes:copy()` should do it

**Reply by danielo515** [reply]
I'm getting this error when doing copy:

```

Reloaded which keys
"danielo_nvim/src/vim/Vim.hx" 70L, 1755B
E5108: Error executing lua ...ack/packer/opt/persistence.nvim/lua/persistence/init.lua:56: Vim(lua):E5108: Error executing lua ...te/pack/packer/start/filetype.nvim/lua/filetype/init.lua:8: Vim(lua):E5108: Error executing lua ...ck/packer/start/LuaSnip/lua/luasnip/loaders/from_lua.lua:41: Failed to load /Users/danielo/.config/lvim/luasnippets/haxe.lua
stack traceback:
	[C]: in function 'error'
	...ck/packer/start/LuaSnip/lua/luasnip/loaders/from_lua.lua:41: in function 'load_files'
	...ck/packer/start/LuaSnip/lua/luasnip/loaders/from_lua.lua:81: in function '_load_lazy_loaded_ft'
	...ck/packer/start/LuaSnip/lua/luasnip/loaders/from_lua.lua:90: in function '_load_lazy_loaded'
	[string ":lua"]:1: in main chunk
	[C]: in function '__newindex'
	...te/pack/packer/start/filetype.nvim/lua/filetype/init.lua:8: in function 'setf'
	...te/pack/packer/start/filetype.nvim/lua/filetype/init.lua:14: in function 'set_filetype'
	...te/pack/packer/start/filetype.nvim/lua/filetype/init.lua:65: in function 'try_lookup'
	...te/pack/packer/start/filetype.nvim/lua/filetype/init.lua:158: in function 'resolve'
	[string ":lua"]:1: in main chunk
	[C]: in function 'cmd'
	...ack/packer/opt/persistence.nvim/lua/persistence/init.lua:56: in function 'load'
	[string ":lua"]:1: in main chunk
stack traceback:
	[C]: in function '__newindex'
	...te/pack/packer/start/filetype.nvim/lua/filetype/init.lua:8: in function 'setf'
	...te/pack/packer/start/filetype.nvim/lua/filetype/init.lua:14: in function 'set_filetype'
	...te/pack/packer/start/filetype.nvim/lua/filetype/init.lua:65: in function 'try_lookup'
	...te/pack/packer/start/filetype.nvim/lua/filetype/init.lua:158: in function 'resolve'
	[string ":lua"]:1: in main chunk
	[C]: in function 'cmd'
	...ack/packer/opt/persistence.nvim/lua/persistence/init.lua:56: in function 'load'
	[string ":lua"]:1: in main chunk
stack traceback:
	[C]: in function 'cmd'
	...ack/packer/opt/persistence.nvim/lua/persistence/init.lua:56: in function 'load'
	[string ":lua"]:1: in main chunk
17:00:50 [DEBUG] lvim: "Activated mac keymappings"    file="keymappings.lua", line=104
17:00:50 [DEBUG] lvim: "loading plugins configuration" file="plugin-loader.lua", line=99
17:00:50 [DEBUG] lvim: "Starting LunarVim"            file="init.lua", line=22
17:00:50 [DEBUG] lvim: "Setting up LSP support"       file="init.lua", line=90
17:00:50 [DEBUG] lvim: "enabled format-on-save"       file="autocmds.lua", line=130
17:00:50 [DEBUG] lvim: "mason configuration not found for haxe_language_server" file="manager.lua", line=11
17:01:12 [DEBUG] lvim: "Called lsp.on_attach_callback" file="init.lua", line=66
17:01:12 [DEBUG] lvim: "skipping setup for document_symbols, method not supported by null-ls" file="utils.lua", line=115
17:01:12 [DEBUG] lvim: "Called lsp.on_attach_callback" file="init.lua", line=66
"danielo_nvim/src/vim/Vim.hx" [Read errors] line 1 of 70 --1%-- col 1
"[Prompt]" [+] line 1 of 1 --100%-- col 1
```

**Reply by L3MON4D3** [reply]
Ah, my bad🤦
_nodes is just a raw list/table of snippets, copy is implemented on snippetNodes.
I think it's easiest to wrap them before doing the copy, not sure if that should be inside the function, or if it should receive a snoppetNode

**Reply by danielo515** [reply]
Do you have a suggested implementation? I'm afraid I am a bit stuck

**Reply by L3MON4D3** [reply]
Here, this works for me:
```lua
ls.setup_snip_env()

local rec_ls
rec_ls = function(pos, _nodes)
  assert(type(_nodes) == "table", "List of nodes must be a table")
  -- flat-copy nodes, otherwise we'd insert `d(...)` in every iteration.
  local nodes = {unpack(_nodes)}
  return function()
    table.insert(nodes, d(pos, rec_ls(pos, _nodes), {}))
    return sn(nil, {
      c(1, {
        -- important!! Having the sn(...) as the first choice will cause infinite recursion.
        t { "" },
        -- deep copy of nodes here.
        sn(nil, nodes):copy(),
      }),
    })
  end
end

ls.snip_expand(s(
    { trig = "ext", dscr = "Extern class", regTrig = false },

    fmt(
      [[ extern class {} {{
              {}
              {}
              }}
    ]],
      {
        i(1),
        d(2, rec_ls(3, fmt(" static function {}({}):{}; ", { i(1), i(2), i(0) })), {}),
        i(3)
      }
    )
  ))
```
This could be improved by appending the `d(...)` only once, and then reusing that table. I guess you'd need another function to make that work.
But also, I'd like to recommend the different approach again, it will result in a more solid snippet imo :sweat_smile:

**Reply by danielo515** [reply]
Thank you for looking into it.
That at least "compiles", but then I get this error when expanding the option choice

```
E5108: Error executing lua: ...ite/pack/packer/start/LuaSnip/lua/luasnip/nodes/util.lua:24: attempt to get length of local 'position_so_far' (a nil value)
stack traceback:
	...ite/pack/packer/start/LuaSnip/lua/luasnip/nodes/util.lua:24: in function 'init_insert_positions'
	...k/packer/start/LuaSnip/lua/luasnip/nodes/dynamicNode.lua:161: in function 'update'
	...k/packer/start/LuaSnip/lua/luasnip/nodes/dynamicNode.lua:342: in function 'update_restore'
	.../pack/packer/start/LuaSnip/lua/luasnip/nodes/snippet.lua:825: in function 'update_restore'
	...ck/packer/start/LuaSnip/lua/luasnip/nodes/choiceNode.lua:254: in function 'no_region_check_wrap'
	...rvim/site/pack/packer/start/LuaSnip/lua/luasnip/init.lua:361: in function 'change_choice'
	/Users/danielo/.config/lvim/lua/user/plugins/cmp-config.lua:177: in function 'on_keymap'
	...unarvim/site/pack/packer/start/nvim-cmp/lua/cmp/core.lua:145: in function 'callback'
	...site/pack/packer/start/nvim-cmp/lua/cmp/utils/keymap.lua:133: in function <...site/pack/packer/start/nvim-cmp/lua/cmp/utils/keymap.lua:127>
E5108: Error executing lua: ...site/pack/packer/start/LuaSnip/lua/luasnip/util/util.lua:505: attempt to perform arithmetic on a nil value
stack traceback:
	...site/pack/packer/start/LuaSnip/lua/luasnip/util/util.lua:505: in function 'pos_sub'
	...ck/packer/start/LuaSnip/lua/luasnip/nodes/choiceNode.lua:223: in function 'no_region_check_wrap'
	...rvim/site/pack/packer/start/LuaSnip/lua/luasnip/init.lua:361: in function 'change_choice'
	/Users/danielo/.config/lvim/lua/user/plugins/cmp-config.lua:177: in function 'on_keymap'
	...unarvim/site/pack/packer/start/nvim-cmp/lua/cmp/core.lua:145: in function 'callback'
	...site/pack/packer/start/nvim-cmp/lua/cmp/utils/keymap.lua:133: in function <...site/pack/packer/start/nvim-cmp/lua/cmp/utils/keymap.lua:127>
```

Maybe I need a newer version of LuaSnip?
> But also, I'd like to recommend the different approach again, it will result in a more solid snippet imo 😅

I took a look at it, and it seems to require some global keybindings. Something that I hate about neovim and that I'm trying to avoid or at least reduce, is that the editor does not have any sense of context, you have to keep it in your head. I'm on a snippet? Then bindigs A to B. The snippet is a choice snippet? Then bindings C to E and if this is an ever expanding snippet, then bindings X to Z. Apply that to all the possible plugins and situations you have in your daily coding and you get an explosion of things to remember. I prefer the editor to remember those and do different things based on the context. If you have any method to reduce this mental burden I will be more than happy to try it.

**Reply by L3MON4D3** [reply]
> Maybe I need a newer version of LuaSnip?

Mhm, I don't think that's it, we didn't change anything there for quite a while.
Very weird.. could you try `:luafile %`ing the code I sent?

**Reply by L3MON4D3** [reply]
> I took a look at it, and it seems to require some global keybindings. Something that I hate about neovim and that I'm trying to avoid or at least reduce, is that the editor does not have any sense of context, you have to keep it in your head. I'm on a snippet? Then bindigs A to B. The snippet is a choice snippet? Then bindings C to E and if this is an ever expanding snippet, then bindings X to Z. Apply that to all the possible plugins and situations you have in your daily coding and you get an explosion of things to remember. I prefer the editor to remember those and do different things based on the context. If you have any method to reduce this mental burden I will be more than happy to try it.

I can see your point regarding mental overhead, I guess the only solution is getting used to it :/

**Comment by danielo515** [comment]
Ah, OK. I didn't know that clone performs a deep clone. Is that a feature
of luasnip? Or is plain Lua?

**Reply by L3MON4D3** [reply]
Feature of luasnip, but all that's happening behind the scenes is `vim.deepcopy()` (though you'd better use the `copy`-method, it's more future-proof in case we have to make changes there)
